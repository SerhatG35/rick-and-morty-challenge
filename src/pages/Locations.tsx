/* eslint-disable react-hooks/exhaustive-deps */
import { LocationTypes } from 'global'
import { useEffect, useState } from 'react'
import { useInView } from 'react-intersection-observer'
import LocationBox from 'src/components/LocationBox'
import { LocationsGET } from 'src/service/dataFetch'
import { Center } from '../styles/styles'

const Locations = () => {
    const [locations, setLocations] = useState<LocationTypes[]>([])
    const [page, setPage] = useState<number | null>(1)
    const [ref, inView] = useInView()

    const fetchLocations = async () => {
        if (page !== null) {
            const data = await LocationsGET(page)
            setLocations(prev => [...prev, ...data.results])
            if (data.info.next !== null) setPage(Number(data.info.next.split('page=')[1]))
            else setPage(data.info.next)
        } else return
    }

    useEffect(() => {
        fetchLocations()
    }, [])

    useEffect(() => {
        if (inView === true) fetchLocations()
    }, [inView])

    return (
        <Center as='main' justifyContent='flex-start' width='70%'>
            {locations?.map((location, index) => (
                <LocationBox
                    key={location.id}
                    location={location}
                    viewRef={index === locations.length - 1 ? ref : undefined}
                />
            ))}
        </Center>
    )
}

export default Locations
