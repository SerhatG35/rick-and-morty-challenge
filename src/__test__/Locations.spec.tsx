import { render, waitFor } from '@testing-library/react'
import { useInView } from 'react-intersection-observer'
import Locations from 'src/pages/Locations'

const useInViewMock = useInView as jest.Mock

describe('Locations page tests', () => {
    test('generates a snapshot', async () => {
        useInViewMock.mockImplementation(() => [null, true])

        const wrapper = render(<Locations />)
        await waitFor(() => wrapper)

        expect(wrapper).toMatchSnapshot()
        wrapper.unmount()
    })
})
