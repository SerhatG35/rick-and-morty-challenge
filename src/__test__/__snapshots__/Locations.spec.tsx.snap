// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Locations page tests generates a snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="sc-bdnxRM dktVZV"
        width="70%"
      >
        <section
          class="sc-bdnxRM byZzmS"
          data-testid="location-wrapper"
        >
          <h1
            class="sc-jSFjdj cYWCdr"
          >
            #75-Earth (C-137)
          </h1>
          <p
            class="sc-iCoGMd fFHScY"
          >
            This is a Planet located in Dimension C-137. There are total of 1 known residents that are originated from here
          </p>
          <div
            class="sc-bdnxRM dNzzIz"
          >
            <article
              class="sc-bdnxRM jDhRBX"
              scale="1.02"
              width="30%"
            >
              <span
                class=" lazy-load-image-background blur"
                style="display: inline-block; height: 50%; width: 100%;"
              >
                <img
                  alt="episode character"
                  height="50%"
                  src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
                  width="100%"
                />
              </span>
              <h2
                class="sc-eCApnc hQwxcN"
                color="#1f1f1f"
                font-style="normal"
              >
                Rick Sanchez
              </h2>
              <div
                class="sc-bdnxRM DhlHX"
              >
                <p
                  class="sc-iCoGMd gGpfF"
                >
                  From Earth (C-137), identifies as Male of Human race.
                </p>
                <p
                  class="sc-iCoGMd KOcH"
                >
                  Current Status: Alive
                </p>
              </div>
            </article>
          </div>
        </section>
        <section
          class="sc-bdnxRM byZzmS"
          data-testid="location-wrapper"
        >
          <h1
            class="sc-jSFjdj cYWCdr"
          >
            #75-Earth (C-137)
          </h1>
          <p
            class="sc-iCoGMd fFHScY"
          >
            This is a Planet located in Dimension C-137. There are total of 1 known residents that are originated from here
          </p>
          <div
            class="sc-bdnxRM dNzzIz"
          >
            <article
              class="sc-bdnxRM jDhRBX"
              scale="1.02"
              width="30%"
            >
              <span
                class=" lazy-load-image-background blur"
                style="display: inline-block; height: 50%; width: 100%;"
              >
                <img
                  alt="episode character"
                  height="50%"
                  src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
                  width="100%"
                />
              </span>
              <h2
                class="sc-eCApnc hQwxcN"
                color="#1f1f1f"
                font-style="normal"
              >
                Rick Sanchez
              </h2>
              <div
                class="sc-bdnxRM DhlHX"
              >
                <p
                  class="sc-iCoGMd gGpfF"
                >
                  From Earth (C-137), identifies as Male of Human race.
                </p>
                <p
                  class="sc-iCoGMd KOcH"
                >
                  Current Status: Alive
                </p>
              </div>
            </article>
          </div>
        </section>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="sc-bdnxRM dktVZV"
      width="70%"
    >
      <section
        class="sc-bdnxRM byZzmS"
        data-testid="location-wrapper"
      >
        <h1
          class="sc-jSFjdj cYWCdr"
        >
          #75-Earth (C-137)
        </h1>
        <p
          class="sc-iCoGMd fFHScY"
        >
          This is a Planet located in Dimension C-137. There are total of 1 known residents that are originated from here
        </p>
        <div
          class="sc-bdnxRM dNzzIz"
        >
          <article
            class="sc-bdnxRM jDhRBX"
            scale="1.02"
            width="30%"
          >
            <span
              class=" lazy-load-image-background blur"
              style="display: inline-block; height: 50%; width: 100%;"
            >
              <img
                alt="episode character"
                height="50%"
                src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
                width="100%"
              />
            </span>
            <h2
              class="sc-eCApnc hQwxcN"
              color="#1f1f1f"
              font-style="normal"
            >
              Rick Sanchez
            </h2>
            <div
              class="sc-bdnxRM DhlHX"
            >
              <p
                class="sc-iCoGMd gGpfF"
              >
                From Earth (C-137), identifies as Male of Human race.
              </p>
              <p
                class="sc-iCoGMd KOcH"
              >
                Current Status: Alive
              </p>
            </div>
          </article>
        </div>
      </section>
      <section
        class="sc-bdnxRM byZzmS"
        data-testid="location-wrapper"
      >
        <h1
          class="sc-jSFjdj cYWCdr"
        >
          #75-Earth (C-137)
        </h1>
        <p
          class="sc-iCoGMd fFHScY"
        >
          This is a Planet located in Dimension C-137. There are total of 1 known residents that are originated from here
        </p>
        <div
          class="sc-bdnxRM dNzzIz"
        >
          <article
            class="sc-bdnxRM jDhRBX"
            scale="1.02"
            width="30%"
          >
            <span
              class=" lazy-load-image-background blur"
              style="display: inline-block; height: 50%; width: 100%;"
            >
              <img
                alt="episode character"
                height="50%"
                src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
                width="100%"
              />
            </span>
            <h2
              class="sc-eCApnc hQwxcN"
              color="#1f1f1f"
              font-style="normal"
            >
              Rick Sanchez
            </h2>
            <div
              class="sc-bdnxRM DhlHX"
            >
              <p
                class="sc-iCoGMd gGpfF"
              >
                From Earth (C-137), identifies as Male of Human race.
              </p>
              <p
                class="sc-iCoGMd KOcH"
              >
                Current Status: Alive
              </p>
            </div>
          </article>
        </div>
      </section>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
